proxy/reverse.go

```diff
@@ -203,3 +221,37 @@ func maybeSetForwardedFor(req *http.Request) {
 	}
 	req.Header.Set("X-Forwarded-For", clientIP)
 }
+
+// checkProxyLoop returns true when it detects a proxy loop
+// (when the proxyID has been seen before), and updates the
+// endpoints of reverseProxy removing the cyclic proxy URLs.
+func (p *reverseProxy) checkProxyLoop(req *http.Request) bool {
+	// check if proxyID has been seen before
+	proxyID := p.proxyID
+	seen := false
+	if v, ok := req.Header["X-Etcd-Proxy-Hop-Seen-By"]; ok {
+		for _, id := range strings.Split(strings.Join(v, ", "), ", ") {
+			if id == proxyID {
+				seen = true
+				break
+			}
+		}
+		// fold multiple proxyIDs into one header
+		proxyID = strings.Join(v, ", ") + ", " + proxyID
+	}
+	req.Header.Set("X-Etcd-Proxy-Hop-Seen-By", proxyID)
+
+	// if there is a proxy loop, update endpoints.
+	if seen {
+		p.Lock()
+		p.director.banned = p.proxyURLs
```

Comment:
This is totally wrong.

When ServeHTTP, it asks director to get current endpoint list, and tries to redirect to them one by one. The endpoint list in director is refreshed every some interval to make it sync with the advertised client urls in the cluster, which consists of advertised client urls of all members. When you ban proxy loop, you should only ban the one that leads to the loop.

My thought is to record the url used to proxy in `X-Etcd-Proxy-Hop-Seen-By`, so we could know which endpoint should be banned.